internal/client/consensus/grandpa/warp_sync.go

```diff
@@ -0,0 +1,180 @@
+}
+
+func (w *WarpSyncProof) lastProofBlockNumber() uint64 {
+	if len(w.Proofs) == 0 {
+		return 0
+	}
+	return w.Proofs[len(w.Proofs)-1].Justification.Justification.Commit.TargetNumber + 1
+}
+
+type WarpSyncProofProvider struct {
+	blockState   BlockState
+	grandpaState GrandpaState
+}
+
+// Generate build a warp sync encoded proof starting from the given block hash
+func (np *WarpSyncProofProvider) Generate(start common.Hash) ([]byte, error) {
+	// Get and traverse all GRANDPA authorities changes from the given block hash
+	beginBlockHeader, err := np.blockState.GetHeader(start)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errMissingStartBlock, err)
+	}
+
+	bestLastBlockHeader, err := np.blockState.BestBlockHeader()
+	if err != nil {
+		return nil, fmt.Errorf("getting best block header: %w", err)
+	}
+
+	if beginBlockHeader.Number > bestLastBlockHeader.Number {
+		return nil, errStartBlockNotFinalized
+	}
+
+	authoritySetChanges, err := np.grandpaState.GetAuthoritesChangesFromBlock(beginBlockHeader.Number)
+	if err != nil {
+		return nil, err
+	}
+
+	limitReached := false
+	finalProof := NewWarpSyncProof()
+	for _, blockNumber := range authoritySetChanges {
+		header, err := np.blockState.GetHeaderByNumber(blockNumber)
+		if err != nil {
+			return nil, err
+		}
+
+		encJustification, err := np.blockState.GetJustification(header.Hash()) // get the justification of such block
+		if err != nil {
+			return nil, err
+		}
+
+		justification, err := decodeJustification[hash.H256, uint64, runtime.BlakeTwo256](encJustification)
+		if err != nil {
+			return nil, err
+		}
+
+		fragment := WarpSyncFragment{Header: *header, Justification: *justification}
+
+		// check the proof size
+		limitReached, err = finalProof.addFragment(fragment)
+		if err != nil {
+			return nil, err
+		}
+
+		if limitReached {
+			break
+		}
+	}
+
+	// If the limit is not reached then retrieve the latest (best) justification
+	// and append in the proofs
+	if !limitReached {
+		// the existing best justification must be for a block higher than the
+		// last authority set change. if we didn't prove any authority set
+		// change then we fallback to make sure it's higher or equal to the
+		// initial warp sync block.
+		bestLastBlockHeader, err := np.blockState.BestBlockHeader()
+		if err != nil {
+			return nil, fmt.Errorf("getting best block header: %w", err)
+		}
+		latestJustification, err := np.blockState.GetJustification(bestLastBlockHeader.Hash())
+		if err != nil {
+			return nil, err
+		}
+
+		justification, err := decodeJustification[hash.H256, uint64, runtime.BlakeTwo256](latestJustification)
+		if err != nil {
+			return nil, err
+		}
+
+		if justification.Justification.Commit.TargetNumber >= finalProof.lastProofBlockNumber() {
+			fragment := WarpSyncFragment{Header: *bestLastBlockHeader, Justification: *justification}
+			_, err = finalProof.addFragment(fragment)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		finalProof.IsFinished = true
```

Comment:
If we dont reach the limit, is there a chance that there are auth set changes between where we stopped and the most recent justification? Im not sure why this would happen, but I'm a bit confused about this "if !limit reached add the highest block number justification" logic

Reply:
There is no problem if there are "missing" auth set changes, since we are sending the best proof we can build